### PR TITLE
crypto: rework CryptoProvider as struct

### PR DESCRIPTION
**Pull request by cpu**
This branch replaces the existing `CryptoProvider` trait with a `CryptoProvider` struct. This has several advantages:

* it consolidates all of the cryptography related settings into one API surface, the `CryptoProvider` struct members. Previously the provider had methods to suggest default ciphersuites, key exchanges etc, but the builder API methods could override them in confusing ways.
* it allows removing the `WantsCipherSuites` and `WantsKxGroups` builder states - the "safe defaults" are automatically supplied by the choice of a crypto provider. Customization is achieved by overriding the provider's struct fields. Having fewer builder states makes the API easier to understand and document.
* it makes customization easier: the end user can rely on ["struct update syntax"](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) to only specify fields values for the required
  customization, and defer the rest to an existing `CryptoProvider`. 
  
Achieving this requires a couple of additional changes:

* The `fill_random` and `load_private_key` fns of the `CryptoProvider` trait are broken out into separate traits, `SecureRandom` and `KeyProvider`, so that we can hold separate `&dyn` references to them in the new `CryptoProvider` struct.
* The cipher suite and key exchange groups are now expressed as `Vec` elements. This avoids imposing a `&'static` lifetime that would preclude runtime customization (e.g. the tls*-mio examples that build the list of ciphersuites at runtime based on command line flags).
* As a result of the `Vec` members we can no longer offer the concrete `CryptoProvider`s as `static` members of their respective modules. Instead we add `pub fn provider() -> CryptoProvider` methods to the `ring` and `aws-lc-rs` module that construct the `CryptoProvider` with the safe defaults, ready for further customization.

Finally, to avoid an unnecessary `Result` unwrap when using the default crypto provider and default protocol versions, or customized protocol versions, the `ClientConfig` and `ServerConfig` default `builder` fns transition directly to `WantsVerifier` using default safe protocol versions. Customization of only protocol versions can be done with new `builder_with_protocol_versions` fns that transition to `WantsVerifier` without needing an unwrap. Using a custom provider requires handling the `Result` that may occur if either the default safe protocol versions or provided custom versions can't be supported by the provider's ciphersuites.

Updates #1372, #1603

### CODE REVIEW

File: examples/src/bin/server_acceptor.rs [hunk 1]
```diff
@@ -225,7 +225,8 @@ impl TestPki {
         // based on the ClientHello (e.g. selecting a different certificate, or customizing
         // supported algorithms/protocol versions).
         let mut server_config = ServerConfig::builder()
-            .with_safe_defaults()
+            .with_safe_default_protocol_versions()
+            .unwrap()
             .with_client_cert_verifier(verifier)
             .with_single_cert(
                 vec![self.server_cert_der.clone()],
```
Comment (cpu) [inline examples/src/bin/server_acceptor.rs:232]: I think one downside of this branch is that the paved-road where you create a builder with the default crypto provider, and then use `with_safe_default_protocol_versions()` to use default protocol versions requires an `unwrap` before continuing the build.

In the case where the default provider is used we have confidence there won't be an error result: the default ciphersuites from *ring* and aws-lc-rs are compatible with all protocol versions. The potential for error only exists when `builder_with_provider()` is used.

Perhaps a more ergonomic solution is to have:

* `ServerConfig::builder()` -> builds a server config with the default crypto provider and the safe protocol defaults, transitions to `WantsVerifier` state without needing an unwrap.
* `ServerConfig::builder_with_protocol_versions()` -> builds a server config with the default crypto provider, and a specified set of protocol versions, transitions to `WantsVerifier` without needing an unwrap.
* `ServerConfig::builder_with_provider()` -> builds a server config with a specified provider, transitions to `WantsVersions` where both `with_safe_default_protocol_versions()` and `with_protocol_versions()` allow customization of protocol versions, but requires handling a `Result` to get to `WantsVerifier`

In general this relates to the "Multiple Unwraps" problem called out in https://github.com/rustls/rustls/issues/1372

Comment (cpu) [reply examples/src/bin/server_acceptor.rs]: I implemented this approach in [572f22b](https://github.com/rustls/rustls/pull/1628/commits/572f22b7b433485eeb9cb0454857a709d7fa9bbb) and I think it's an improvement.

If preferred I could break that out into a separate PR. In the meantime I'll resolve this thread.

File: rustls/examples/internal/bogo_shim_impl.rs [hunk 2]
```diff
@@ -19,9 +19,9 @@ use rustls::{
 };
 
 #[cfg(all(not(feature = "ring"), feature = "aws_lc_rs"))]
-use rustls::crypto::{aws_lc_rs as provider, aws_lc_rs::AWS_LC_RS as PROVIDER};
+use rustls::crypto::aws_lc_rs as primary_provider;
```
Comment (djc) [inline rustls/examples/internal/bogo_shim_impl.rs:22]: I think I'd just name this `provider`? `primary_provider` seems like a weird notion since I don't think there is another provider in play in the bogo shim at the same time? I guess you're trying to disambiguate from the `provider()` function but functions and modules live in a different namespace so `provider::provider()` should Just Work IIRC.

File: rustls/src/client/client_conn.rs [hunk 3]
```diff
@@ -226,9 +216,7 @@ pub enum Tls12Resumption {
 impl Clone for ClientConfig {
     fn clone(&self) -> Self {
         Self {
-            cipher_suites: self.cipher_suites.clone(),
-            kx_groups: self.kx_groups.clone(),
-            provider: self.provider,
+            provider: Arc::<CryptoProvider>::clone(&self.provider),
```
Comment (djc) [inline rustls/src/client/client_conn.rs:221]: Do we really need the turbofish here? That seems surprising to me. If so maybe just prefer `self.provider.clone()` instead?

File: rustls/src/crypto/mod.rs [hunk 4]
```diff
@@ -149,39 +140,33 @@ pub use crate::msgs::handshake::KeyExchangeAlgorithm;
...
     /// These are used for both certificate chain verification and handshake signature verification.
-    ///
-    /// This is called by [`crate::ConfigBuilder::with_root_certificates()`],
-    /// [`crate::server::WebPkiClientVerifier::builder_with_provider()`] and
-    /// [`crate::client::WebPkiServerVerifier::builder_with_provider()`].
-    fn signature_verification_algorithms(&self) -> WebPkiSupportedAlgorithms;
+    pub signature_verification_algorithms: WebPkiSupportedAlgorithms,
+
+    /// Source of cryptographically secure random numbers.
+    pub secure_random: &'static dyn SecureRandom,
```
Comment (djc) [inline rustls/src/crypto/mod.rs:166]: Might be nice to experiment with changing this to be a separate commit? That is, I think the change from `CryptoProvider` going from is-a `SecureRandom` to `has-a` `SecureRandom` could I think be orthogonal to `CryptoProvider` going from `trait` to `struct` -- it could return the `&'static dyn SecureRandom` in a method. (Same for `KeyProvider`, of course.)

File: rustls/src/crypto/ring/mod.rs [hunk 5]
```diff
@@ -29,26 +29,20 @@ pub(crate) mod tls13;
 /// A `CryptoProvider` backed by the [*ring*] crate.
 ///
 /// [*ring*]: https://github.com/briansmith/ring
-pub static RING: &dyn CryptoProvider = &Ring;
+pub fn provider() -> CryptoProvider {
```
Comment (djc) [inline rustls/src/crypto/ring/mod.rs:32]: I wonder if we should give this a more opinionated name? Of course this is not the only possible `CryptoProvider` instantiation based on *ring* cryptography, merely the one with default cipher suites, key exchange groups and signature algorithms. Maybe `default_provider()`?

**Review by djc — APPROVED** [review]
I think this is looking pretty good.

File: rustls/examples/internal/bogo_shim_impl.rs (hunk 2 above)
Comment (cpu) [reply rustls/examples/internal/bogo_shim_impl.rs:22]: > I guess you're trying to disambiguate from the provider() function

I think I picked this up from another module that was using `primary_provider` and thought it did a good job distinguishing from the provider() function, but as you mention maybe that's not necessary. 

I'll adjust to just use `primary` throughout :+1:

File: rustls/src/client/client_conn.rs (hunk 3 above)
Comment (cpu) [reply rustls/src/client/client_conn.rs:221]: IIRC `clippy` was flagging this without the turbofish but I will verify.

File: rustls/src/crypto/mod.rs (hunk 4 above)
Comment (cpu) [reply rustls/src/crypto/mod.rs:166]: > That is, I think the change from CryptoProvider going from is-a SecureRandom to has-a SecureRandom could I think be orthogonal to CryptoProvider going from trait to struct

Sure, that makes sense :+1: I will split it out.

File: rustls/src/crypto/ring/mod.rs (hunk 5 above)
Comment (cpu) [reply rustls/src/crypto/ring/mod.rs:32]: `default_provider` sounds good to me. I agree it would be nice to emphasize the intent is that this is a starting point for sensible defaults that can be further customized.

File: rustls/src/client/client_conn.rs (hunk 3 above)
Comment (cpu) [reply rustls/src/client/client_conn.rs:221]: Yeah:
```
error: using `.clone()` on a ref-counted pointer
   --> rustls/src/client/client_conn.rs:221:23
    |
221 |             provider: self.provider.clone(),
    |                       ^^^^^^^^^^^^^^^^^^^^^ help: try: `Arc::<crypto::CryptoProvider>::clone(&self.provider)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr
note: the lint level is defined here
   --> rustls/src/lib.rs:278:5
    |
278 |     clippy::clone_on_ref_ptr,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `rustls` (lib) due to previous error
```

File: rustls/src/crypto/ring/mod.rs (hunk 5 above)
Comment (briansmith) [reply rustls/src/crypto/ring/mod.rs:32]: > `default_provider` sounds good to me. I agree it would be nice to emphasize the intent is that this is a starting point for sensible defaults that can be further customized.

I don't object to that. I expect that some packagers of Rustls (e.g. Linux distros) will change the default provider to something that isn't *ring*, like OpenSSL Or Windows CNG. And then people who want/need to use *ring* (e.g. because of FIPS, in the future) will need a way to force the use of *ring*. So ideally there would be both a way to force the use of *ring* and a `default_provider()`.

For the same reason, enabling "FIPS mode" shouldn't affect which provider is used; the option should only change how the provider is configured. So, for example if one wants to use AWS-LC in FIPS mode, they should choose AWS-LC *and* also choose "FIPS mode."

File: rustls/examples/internal/bogo_shim_impl.rs (hunk 2 above)
Comment (cpu) [reply rustls/examples/internal/bogo_shim_impl.rs]: Fixed.

File: rustls/src/crypto/mod.rs (hunk 4 above)
Comment (cpu) [reply rustls/src/crypto/mod.rs:170]: I split out two additional commits for this, one for each new trait.

File: rustls/src/crypto/ring/mod.rs (hunk 5 above)
Comment (cpu) [reply rustls/src/crypto/ring/mod.rs]: I've switched to `default_provider` for the constructors.

>  So ideally there would be both a way to force the use of ring and a default_provider().

That's certainly possible. Using `ClientConfig::builder` or `ServerConfig::builder` will default to *ring*, but if you wanted to make the choice explicit it's just sugar over `builder_with_provider(rustls::crypto::ring::default_provider().into())`.

File: rustls/src/client/client_conn.rs (hunk 3 above)
Comment (djc) [reply rustls/src/client/client_conn.rs:221]: Yeah... I don't care much for that particular lint, but okay.

File: rustls/src/builder.rs [hunk 6]
```diff
@@ -1,62 +1,55 @@
...
+use std::sync::Arc;
 
 /// A [builder] for [`ServerConfig`] or [`ClientConfig`] values.
 ///
 /// To get one of these, call [`ServerConfig::builder()`] or [`ClientConfig::builder()`].
 ///
 /// To build a config, you must make at least three decisions (in order):
 ///
-/// - Which protocol primitives should be supported (cipher suites, key exchange groups, protocol versions)?
+/// - Which cryptographic provider to use, such as [`ring::provider`]?
```
Comment (jsha) [inline rustls/src/builder.rs:16]: The new API reduces the number of "must decides." In particular the choice of protocol primitives went away. Now I think it's more accurate to say:

```
/// To build a config, you must make at least two decisions (in order):
///
/// - How should this client or server verify certificates provided by its peer?
/// - What certificates should this client or server present to its peer?
```

To some extent you're "making a decision" by choosing `ClientConfig::builder` vs `ClientConfig::builder_with_XXX`. But that doesn't have to be part of this list. The goal of this list is to orient people to "what are the must-call methods I use to get to the point where I can finish building a config."

**Review by jsha — COMMENTED** [review]
Thanks for working on this! Looks great.

The `unwrap` for checking cipher suite / TLS version compatibility still feels awkward, even with introducing the `builder_with_protocol_versions`.

There was some prior discussion when the ConfigBuilder was first introduced, starting here: https://github.com/rustls/rustls/pull/564#issuecomment-826049679

The suggestion was to have different types for TLS 1.2 cipher suites vs TLS 1.3 cipher suites. In fact we have that already, we just wrap them in an enum so they can share a type:

```rust
pub enum SupportedCipherSuite {
    Tls12(&'static Tls12CipherSuite),
    Tls13(&'static Tls13CipherSuite),
}
```

What if `CryptoProvider` had two fields:

```rust
tls12_cipher_suites: Vec<&'static Tls12CipherSuite>,
tls13_cipher_suites: Vec<&'static Tls13CipherSuite>,
```

And the way to disable TLS 1.2 would be to provide an empty `Vec` for `tls12_cipher_suites`. That way we could do away with both the `unwrap()` and `builder_with_protocol_versions`.

Under the hood, the builder could then bundle those up into the enum `SupportedCipherSuite` to minimize unnecessary diffs.

**Comment by djc** [comment]
> And the way to disable TLS 1.2 would be to provide an empty `Vec` for `tls12_cipher_suites`.

This seems like a good idea to me.

I think I advocated at the time that there was no need to specify protocol versions anyway because this was implicitly expressed by the supply of cipher suites. I think that was deemed not explicit enough but just wanted to mention it here for completeness.

But most of this discussion seems out of scope for this specific PR?

File: rustls/tests/api.rs [hunk 7]
```diff
@@ -5563,30 +5589,23 @@ impl rustls::crypto::SecureRandom for FaultyRandom {
...
-    };
 
     let client_config = finish_client_config(
         KeyType::Rsa,
-        rustls::ClientConfig::builder_with_provider(&TEST_PROVIDER).with_safe_defaults(),
+        rustls::ClientConfig::builder_with_provider(
+            CryptoProvider {
+                secure_random: &FAULTY_RANDOM,
+                ..provider::default_provider()
+            }
```
Comment (ctz) [inline rustls/tests/api.rs:5589]: _so_ much nicer!

File: examples/src/bin/server_acceptor.rs [hunk 8]
```diff
@@ -225,8 +225,6 @@ impl TestPki {
         // based on the ClientHello (e.g. selecting a different certificate, or customizing
         // supported algorithms/protocol versions).
         let mut server_config = ServerConfig::builder()
-            .with_safe_default_protocol_versions()
-            .unwrap()
```
Comment (ctz) [inline examples/src/bin/server_acceptor.rs]: non-actionable thinking aloud:

with this commit, i probably would have gone for a different route: making `with_safe_default_protocol_versions()` infallible (it returns `WantsVerifier`), but that bakes-in the current property of that function that it enables both protocol versions (and, therefore, there's no route to achieving the degenerate TLS1.2-version but TLS1.3-suites or TLS1.3-version but TLS1.2-suites combinations that are unusable)

File: rustls/src/client/client_conn.rs [hunk 9]
```diff
@@ -232,17 +234,41 @@ impl Clone for ClientConfig {
...
+    /// [`CryptoProvider`]: [`crate::crypto::ring::default_provider`], safe ciphersuite defaults and
+    /// the provided protocol versions.
     ///
     /// For more information, see the [`ConfigBuilder`] documentation.
-    pub fn builder() -> ConfigBuilder<Self, WantsVersions> {
+    #[cfg(feature = "ring")]
+    pub fn builder_with_protocol_versions(
+        versions: &[&'static versions::SupportedProtocolVersion],
+    ) -> ConfigBuilder<Self, WantsVerifier> {
+        // Safety: we know the *ring* provider's ciphersuites are compatible with all protocol version choices.
```
Comment (ctz) [inline rustls/src/client/client_conn.rs:259]: moreover: we know that the provider has at least one ciphersuite for each version

i think this will panic for `builder_with_protocol_versions(&[])`; i don't think that necessarily needs fixing.

**Review by ctz — APPROVED** [review]
IMO this is a big improvement

File: rustls/src/builder.rs (hunk 6 above)
Comment (cpu) [reply rustls/src/builder.rs]: Fixed up. Thanks!

File: rustls/src/client/client_conn.rs (hunk 9 above)
Comment (cpu) [reply rustls/src/client/client_conn.rs:261]: > i think this will panic for builder_with_protocol_versions(&[]); i don't think that necessarily needs fixing.

I added a note to each fn's docstring about this :+1:

**Comment by ctz** [comment]
> What if `CryptoProvider` had two fields:
> 
> ```rust
> tls12_cipher_suites: Vec<&'static Tls12CipherSuite>,
> tls13_cipher_suites: Vec<&'static Tls13CipherSuite>,
> ```

I guess we would put `tls13_cipher_suites` before `tls12_cipher_suites` on the wire? That seems like a bit of a policy decision that is currently not made by the library. Though -- hopefully -- nobody does version negotiation between TLS1.2 and TLS1.3 incorrectly based on the offered cipher suites?

> And the way to disable TLS 1.2 would be to provide an empty `Vec` for `tls12_cipher_suites`. That way we could do away with both the `unwrap()` and `builder_with_protocol_versions`.

We'd also need to put a `&'static SupportedProtocolVersion` in each `Tls12CipherSuite` and `Tls13CipherSuite`, and then require that everyone fills in the right one? Otherwise #224 would become impossible.

**Comment by jsha** [comment]
> I guess we would put tls13_cipher_suites before tls12_cipher_suites on the wire? That seems like a bit of a policy decision that is currently not made by the library.

This is a good point, though as you say:

> hopefully -- nobody does version negotiation between TLS1.2 and TLS1.3 incorrectly based on the offered cipher suites?

So I think it's an okay policy decision to make.

> We'd also need to put a &'static SupportedProtocolVersion in each Tls12CipherSuite and Tls13CipherSuite, and then require that everyone fills in the right one? Otherwise https://github.com/rustls/rustls/issues/224 would become impossible.

Also a good point, and very achievable I think. But this does raise my estimate of how much work is involved.

**Comment by jsha** [comment]
Filed as a separate issue: #1659
